lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,397 @@
+    else:
+        LOG.debug(f"Target «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "control_status": control_status
+        }
+
+        target = db.fetch_row("""
+            insert into warehouse.target (identifier, control_status)
+                values (%(identifier)s, %(control_status)s)
+            returning target_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created target {target.id} «{target.identifier}»")
+
+    return target
+
+
+def target_control_status(control_status: str, target_identifier: str) -> bool:
+    """
+    Determine the control status of the target.
+    """
+    expected_values = ["NotControl", "PositiveControl"]
+    if not control_status or control_status not in expected_values: 
+        raise UnknownControlStatusError(f"Unknown control status «{control_status}» for target {target_identifier}")
+    return control_status == "PositiveControl"
+
+
+def find_or_create_sample(db: DatabaseSession, identifier: str,
+                          details: dict) -> Any:
+    """
+    Select sample by *identifier*, or insert it if it doesn't exist.
+
+    TODO eventually we'll need to add the encounter_id once the relationship
+    between encounter_id and sample_id exists in the db.
+    """
+    LOG.debug(f"Looking up sample «{identifier}»")
+
+    sample = db.fetch_row("""
+        select sample_id as id, identifier
+          from warehouse.sample
+         where identifier = %s
+        """, (identifier,))
+
+    if sample:
+        LOG.info(f"Found sample {sample.id} «{sample.identifier}»")
+    else:
+        LOG.debug(f"Sample «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "details": Json(details)
+            # "encounter_id"
+        }
+
+        sample = db.fetch_row("""
+            insert into warehouse.sample (
+                identifier, 
+                details)
+                
+                values (
+                    %(identifier)s, 
+                    %(details)s)
+                    
+            returning sample_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created sample {sample.id} «{sample.identifier}»")
+
+    return sample
+
+def sample_identifier(db: DatabaseSession, barcode: str) -> str:
+    """
+    Find corresponding UUID for scanned sample barcode within
+    warehouse.identifier.
+
+    Returns barcode if it starts with "SCH-" as this indicates it 
+    is a control sample, and therefore does not have a UUID.
+    
+    TODO determine course of action if barcode not found
+    within warehouse.identifier.
+    """
+
+    LOG.debug(f"Looking up sample barcode {barcode} to find UUID")
+
+    uuid = db.fetch_row("""
+        select uuid, name
+          from warehouse.identifier
+          join warehouse.identifier_set using (identifier_set_id)
+         where barcode = %s
```

Comment:
As a todo item, we should make sure this is happening case-insensitively. `barcode` is a `citext` type (case-insensitive text), but we need to check that the placeholder value we're passing in on the right-hand side also ends up as `citext` and not just `text`.

Reply:
We checked, and they're compared case-insensitively!